mir/src/passes/unrolling/match_optimizer.rs

```diff
@@ -0,0 +1,241 @@
+
+            // Start picking constraints from the biggest sets that evaluate to the same values
+            for eval_index in eval_lens.values().cloned().rev().flatten() {
+                let constraints = self.constraints_evaluation_indices.get(&eval_index).unwrap();
+
+                if have_disjoint_conditions(new_constraint.clone(), constraints) {
+                    // If the new constraints are disjoint from the current ones, we can add
+                    // them
+                    new_constraint.push(constraints.clone());
+                    taken_eval_indices.push(eval_index);
+                }
+            }
+
+            // Remove the picked evaluation indices from all structures
+            for eval_index in taken_eval_indices {
+                self.constraints_evaluation_indices.remove(&eval_index);
+                eval_lens.iter_mut().for_each(|(_len, evals)| {
+                    evals.retain(|&e_idx| e_idx != eval_index);
+                });
+                eval_lens.retain(|_, constraints| !constraints.is_empty());
+            }
+
+            // Create combined constraint
+            let mut cur_node = None;
+            for equivalent_constraints in new_constraint {
+                let mut cur_condition = None;
+                for (condition, _) in equivalent_constraints.clone() {
+                    if cur_condition.is_none() {
+                        cur_condition = Some(condition.clone());
+                    } else {
+                        cur_condition =
+                            Some(Add::create(cur_condition.unwrap(), condition.clone(), span));
+                    }
+                }
+                let new_node = Mul::create(
+                    cur_condition.unwrap(),
+                    equivalent_constraints.first().unwrap().1.clone(),
+                    span,
+                );
+
+                cur_node = match cur_node {
+                    Some(existing) => Some(Add::create(existing, new_node, span)),
+                    None => Some(new_node),
+                };
+            }
+
+            // Note: This will ensure the resulting constraint is in the form `Sub(x,y)`
+            // representing `enf x = y`
+            let zero_node = Value::create(SpannedMirValue {
+                span: Default::default(),
+                value: MirValue::Constant(ConstantValue::Felt(0)),
+            });
+            // The following unwrap is safe as we always have at least one constraint above
+            let new_node_with_sub_zero = Sub::create(cur_node.unwrap(), zero_node, span);
+
+            new_vec.push(new_node_with_sub_zero);
+        }
+        new_vec
+    }
+
+    /// Adds bus-related constraints to the resulting constraints vector.
+    pub fn gather_all_constraints(
+        bus_related_constraints: &mut [(Link<Op>, Vec<Link<Op>>)],
+        combined_main_constraints: Vec<Link<Op>>,
+        span: SourceSpan,
+    ) -> Vec<Link<Op>> {
```

Comment:
I would also maybe mention that we are computing latches for the constraints here.